File: src/behavioral/iterator2.java

public interface Iterator<T> {
    boolean hasNext();
    T getNext();
    void reset();
}

public class DepthFirstIterator<T> implements Iterator<T> {
    private final Vertex<T> startVertex;
    private Deque<Vertex<T>> stack = new LinkedList<>();

    public DepthFirstIterator(Vertex<T> startVertex) {
        this.startVertex = startVertex;
        stack.push(startVertex);
    }

    @Override
    public boolean hasNext() {
        return !stack.isEmpty();
    }

    @Override
    public T getNext() {
        if (!hasNext()) {
            return null;
        }
        Vertex<T> current = stack.pop();
        if (!current.isVisited()) {
            current.setVisited(true);
            current.getNeighbors().forEach(stack::push);
            return current;
        }
        return getNext();
    }

    @Override
    public void reset() {
        stack.clear();
        stack.push(startVertex);
    }
}

public class BreadthFirstIterator<T> implements Iterator<T> {
    public BreadthFirstIterator() {
        ...
    }

    @Override
    public boolean hasNext() {
        ...
    }

    @Override
    public T getNext() {
        ...
    }

    @Override
    public void reset() {
        ...
    }
}

public static void main(String[] args) {
    Vertex<Integer> v0 = new Vertex<>(0);
    Vertex<Integer> v1 = new Vertex<>(1);
    Vertex<Integer> v2 = new Vertex<>(2);
    Vertex<Integer> v3 = new Vertex<>(3);
    Vertex<Integer> v4 = new Vertex<>(4);
    Vertex<Integer> v5 = new Vertex<>(5);
    Vertex<Integer> v6 = new Vertex<>(6);

    v0.setNeighbors(Arrays.asList(v1, v5, v6));
    v1.setNeighbors(Arrays.asList(v3, v4, v5));
    v4.setNeighbors(Arrays.asList(v2, v6));
    v6.setNeighbors(Arrays.asList(v0));

    DepthFirstIterator<Integer> dfs = new DepthFirstIterator<>(v0);
    while (dfs.hasNext()) {
        System.out.println(dfs.getNext());
    }
}

public interface Tree<T> {
    Iterator<T> createBFSIterator(...);
    Iterator<T> createDFSIterator(...);
}

public class BinarySearchTree<T> implements Tree<T> {
    ...

    @Override
    public Iterator<T> createBFSIterator(...) {
        return new BreadthFirstIterator<>(...);
    }

    @Override
    public Iterator<T> createDFSIterator(...) {
        return new DepthFirstIterator<>(...);
    }
}

public class RedBlackTree<T> implements Tree<T> {
    ...

    @Override
    public Iterator<T> createBFSIterator(...) {
        return new BreadthFirstIterator<>(...);
    }

    @Override
    public Iterator<T> createDFSIterator(...) {
        return new DepthFirstIterator<>(...);
    }
}

File: src/behavioral/iterator2.ts
interface Iterator<T> {
  hasNext(): boolean;
  getNext(): T | null;
  reset(): void;
}

class Vertex<T> {
  private visited = false;
  private neighbors: Vertex<T>[] = [];

  constructor(public val: T) {}

  isVisited(): boolean {
    return this.visited;
  }

  setVisited(visited: boolean): void {
    this.visited = visited;
  }

  getNeighbors(): Vertex<T>[] {
    return this.neighbors;
  }

  setNeighbors(neighbors: Vertex<T>[]): void {
    this.neighbors = neighbors;
  }
}

class DepthFirstIterator<T> implements Iterator<T> {
  private startVertex: Vertex<T>;
  private stack: Vertex<T>[] = [];

  constructor(startVertex: Vertex<T>) {
    this.startVertex = startVertex;
    this.stack.push(startVertex);
  }

  hasNext(): boolean {
    return this.stack.length > 0;
  }

  getNext(): T | null {
    if (!this.hasNext()) {
      return null;
    }

    const current = this.stack.pop()!;
    if (!current.isVisited()) {
      current.setVisited(true);
      // Push neighbors to stack
      current.getNeighbors().forEach((neighbor) => {
        this.stack.push(neighbor);
      });
      return current.val;
    }
    return this.getNext();
  }

  reset(): void {
    this.stack = [this.startVertex];
  }
}

class BreadthFirstIterator<T> implements Iterator<T> {
  private startVertex: Vertex<T>;
  private queue: Vertex<T>[] = [];

  constructor(startVertex: Vertex<T>) {
    this.startVertex = startVertex;
    this.queue.push(startVertex);
  }

  hasNext(): boolean {
    return this.queue.length > 0;
  }

  getNext(): T | null {
    if (!this.hasNext()) {
      return null;
    }

    const current = this.queue.shift()!;
    if (!current.isVisited()) {
      current.setVisited(true);
      // Push neighbors to queue
      current.getNeighbors().forEach((neighbor) => {
        this.queue.push(neighbor);
      });
      return current.val;
    }
    return this.getNext();
  }

  reset(): void {
    this.queue = [this.startVertex];
  }
}

function iteratorFn(): void {
  const v0 = new Vertex<number>(0);
  const v1 = new Vertex<number>(1);
  const v2 = new Vertex<number>(2);
  const v3 = new Vertex<number>(3);
  const v4 = new Vertex<number>(4);
  const v5 = new Vertex<number>(5);
  const v6 = new Vertex<number>(6);

  v0.setNeighbors([v1, v5, v6]);
  v1.setNeighbors([v3, v4, v5]);
  v4.setNeighbors([v2, v6]);
  v6.setNeighbors([v0]);

  console.log("--- Depth First Search ---");
  const dfs = new DepthFirstIterator<number>(v0);
  while (dfs.hasNext()) {
    const next = dfs.getNext();
    if (next) {
      console.log(next);
    }
  }

  // Reset visit status for BFS
  [v0, v1, v2, v3, v4, v5, v6].forEach((v) => v.setVisited(false));

  console.log("--- Breadth First Search ---");
  const bfs = new BreadthFirstIterator<number>(v0);
  while (bfs.hasNext()) {
    const next = bfs.getNext();
    if (next) {
      console.log(next);
    }
  }
}

iteratorFn();

interface Tree<T> {
  createBFSIterator(...params: unknown[]): Iterator<T>;
  createDFSIterator(...params: unknown[]): Iterator<T>;
}

export class BinarySearchTree<T> implements Tree<T> {
  //

  createBFSIterator(param: Vertex<T>): Iterator<T> {
    return new BreadthFirstIterator<T>(param);
  }

  createDFSIterator(param: Vertex<T>): Iterator<T> {
    return new DepthFirstIterator<T>(param);
  }
}

export class RedBlackTree<T> implements Tree<T> {
  //

  createBFSIterator(param: Vertex<T>): Iterator<T> {
    return new BreadthFirstIterator<T>(param);
  }

  createDFSIterator(param: Vertex<T>): Iterator<T> {
    return new DepthFirstIterator<T>(param);
  }
}
